src/test/java/cloud/fogbow/common/plugins/cloudidp/cloudstack/CloudStackIdentityProviderPluginTest.java

```diff
@@ -0,0 +1,287 @@
+		// set up
+		Map<String, String> credentials = null;
+
+		// exercise
+		this.plugin.getCloudUser(credentials);
+	}
+		
+	// test case: When invoking the getCloudUser method with an empty credential
+	// map, it must throw an InvalidParameterException.
+	@Test(expected = InvalidParameterException.class) // verify
+	public void testGetCloudUserWithEmptyCredentialsMap() throws FogbowException {
+		// set up
+		Map<String, String> credentials = new HashMap<String, String>();
+
+		// exercise
+		this.plugin.getCloudUser(credentials);
+	}
+
+	// test case: When invoking the getCloudUser method with a credential map
+	// without a user name, it must throw an InvalidParameterException.
+	@Test(expected = InvalidParameterException.class) // verify
+	public void testGetCloudUserWithoutUsername() throws FogbowException {
+		// set up
+		Map<String, String> credentials = new HashMap<String, String>();
+		credentials.put(CloudStackConstants.Identity.PASSWORD_KEY_JSON, ANY_VALUE);
+		credentials.put(CloudStackConstants.Identity.DOMAIN_KEY_JSON, ANY_VALUE);
+
+		// exercise
+		this.plugin.getCloudUser(credentials);
+	}
+
+	// test case: When invoking the getCloudUser method with a credential map
+	// without a password, it must throw an InvalidParameterException.
+	@Test(expected = InvalidParameterException.class) // verify
+	public void testGetCloudUserWithoutPassword() throws FogbowException {
+		// set up
+		Map<String, String> credentials = new HashMap<String, String>();
+		credentials.put(CloudStackConstants.Identity.USERNAME_KEY_JSON, ANY_VALUE);
+		credentials.put(CloudStackConstants.Identity.DOMAIN_KEY_JSON, ANY_VALUE);
+
+		// exercise
+		this.plugin.getCloudUser(credentials);
+	}
+
+	// test case: When invoking the getCloudUser method with a credential map
+	// without a domain, it must throw an InvalidParameterException.
+	@Test(expected = InvalidParameterException.class) // verify
+	public void testGetCloudUserWithoutDomain() throws FogbowException {
+		// set up
+		Map<String, String> credentials = new HashMap<String, String>();
+		credentials.put(CloudStackConstants.Identity.USERNAME_KEY_JSON, ANY_VALUE);
+		credentials.put(CloudStackConstants.Identity.PASSWORD_KEY_JSON, ANY_VALUE);
+
+		// exercise
+		this.plugin.getCloudUser(credentials);
+	}
+
+	// test case: When invoking the getCloudUser method with a valid credential map,
+	// and the response to the login request in the server returns an HTTP error
+	// code, it must throw a FogbowException.
+	@Test(expected = FogbowException.class)
+	public void testGetCloudUserWithHttpCodeErrorFromLoginResponse() throws FogbowException {
+		// set up
+		Map<String, String> credentials = new HashMap<String, String>();
+		credentials.put(CloudStackConstants.Identity.USERNAME_KEY_JSON, ANY_VALUE);
+		credentials.put(CloudStackConstants.Identity.PASSWORD_KEY_JSON, ANY_VALUE);
+		credentials.put(CloudStackConstants.Identity.DOMAIN_KEY_JSON, ANY_VALUE);
+
```

Comment:
I think we might make this test code simpler if we extract the interaction with the HttpRequestClient into an protected method in the plugin, them mocking its behaviour in the test. what do you think @ANKerD?  In summary, it would be nice to hide this http stuff. I suggest to refactor the plugin in this direction. After that, I resume the review

Reply:
I think it's a nice move going in that direction. The bit of refactoring I'm working right now addresses this concern.

Reply:
Ok, refactoring is done.

Reply:
as the getCloudUser method depends, directly, on the createLoginRequest, authenticate and getCloudStackUser ones, it seems there is no need to mock the doLoginAuthentication method (if we avoid mocking it, we can avoid HTTP details in this test)